torch/quantization/convert_modules.py

```diff
@@ -0,0 +1,193 @@
+        qconfig object already'
+    module.qconfig = None
+    if prefix in qconfig_dict:
+        module.qconfig = qconfig_dict[prefix]
+    else:
+        module.qconfig = qconfig_parent
+
+    for name, child in module.named_children():
+        module_prefix = prefix + '.' + name if prefix else name
+        propogate_qconfig_helper(child, qconfig_dict, module.qconfig, module_prefix)
+
+def propogate_qconfig(module, qconfig_dict):
+    r"""Propogate qconfig through the module hierarchy given a qconfig dict
+
+    Args:
+        module: input module we want to add qconfig
+        qconfig_dict: dictionary that maps from name of submodule to quantization
+                     configuration
+    """
+    propogate_qconfig_helper(module, qconfig_dict)
+
+def _observer_forward_hook(self, input, output):
+    self.observer(output)
+
+def add_observer(module):
+    r"""Transforms a module to a quantized module according to qconfig_dict.
+
+    This function insert observer module to all leaf child module of a given
+    module based on qconfig_dict.
+
+    Args:
+        module: instance of the module we want to transform
+        qconfig_dict: dictionary that maps from name of submodule to quantization
+                     configuration
+        qconfig_parent: quantization config of parent module, we will fallback to
+                       this config when there is no specified config for current
+                       module
+        prefix: corresponding prefix of the current module, used as key in
+                qconfig_dict
+    """
+    # Insert observers only for leaf nodes, note that this observer is for
+    # the output of the module, for input QuantStub will observe them
+    if module.qconfig is not None and len(module._modules) == 0:
+        # observer and hook will be gone after we swap the module
+        module.add_module('observer', module.qconfig.activation())
+        module.register_forward_hook(_observer_forward_hook)
+
+    for name, child in module.named_children():
+        if name != 'observer':
+            add_observer(child)
+
+def get_config_key(name, qconfig_dict):
+    if name in qconfig_dict:
+        return name
+    elif name:
+        parent = '.'.join(name.split('.')[:-1])
+        return get_config_key(parent, qconfig_dict)
+    else:
+        return '' if '' in qconfig_dict else None
+
+class QuantWrapper(nn.Module):
+    def __init__(self, module, qconfig):
+        super(QuantWrapper, self).__init__()
+        self.quant = QuantStub(qconfig)
+        self.dequant = DeQuantStub()
+        self.module = module
+
+    def forward(self, X):
+        X = self.quant(X)
+        X = self.module(X)
+        return self.dequant(X)
+
+def add_quant_dequant(module):
+    r"""Add QuantStub and DeQuantStub module and Wrap the module in QuantWrapper
+    according to qconfig_dict
+    """
+    is_leaf = len([name for (name, _) in module.named_children() if name != 'observer']) == 0
```

Comment:
you can use `any` instead of `len` - it'd stop after first match

Reply:
since order is changed we don't need this anymore, but thanks for the tip